adafruit_tinylora/adafruit_tinylora.py

```diff
@@ -0,0 +1,397 @@
+        # append preamble to packet
+        lora_pkt[0] = const(0x40)
+        lora_pkt[1] = self._ttn_config.device_address[3]
+        lora_pkt[2] = self._ttn_config.device_address[2]
+        lora_pkt[3] = self._ttn_config.device_address[1]
+        lora_pkt[4] = self._ttn_config.device_address[0]
+        lora_pkt[5] = 0
+        lora_pkt[6] = frame_counter & 0x00FF
+        lora_pkt[7] = (frame_counter >> 8) & 0x00FF
+        lora_pkt[8] = 0x01
+        # set length of LoRa packet
+        lora_pkt_len = 9
+        # load encrypted data into lora_pkt
+        for i in range(0, data_length):
+            lora_pkt[lora_pkt_len + i] = enc_data[i]
+        # recalculate packet length
+        lora_pkt_len = lora_pkt_len + data_length
+        # Calculate MIC
+        mic = bytearray(4)
+        mic = aes.calculate_mic(lora_pkt, lora_pkt_len, mic)
+        # load mic in package
+        for i in range(0, 4):
+            lora_pkt[i + lora_pkt_len] = mic[i]
+        # recalculate packet length (add MIC length)
+        lora_pkt_len += 4
+        self.send_packet(lora_pkt, lora_pkt_len)
+
+    def send_packet(self, lora_packet, packet_length):
+        """Sends a LoRa packet using the RFM Module
+          :param bytearray lora_packet: assembled LoRa packet from send_data
+          :param int packet_length: length of LoRa packet to send
+        """
+        # Set RFM to standby
+        self._write_u8(_MODE_STDBY, 0x81)
+        # wait for RFM to enter standby mode
+        time.sleep(0.01)
+        # switch interrupt to txdone
+        self._write_u8(0x40, 0x40)
+        # check for multi-channel configuration
+        if self._channel is None:
+            self._tx_random = randint(0, 7)
+            self._rfm_lsb = self._frequencies[self._tx_random][2]
+            self._rfm_mid = self._frequencies[self._tx_random][1]
+            self._rfm_msb = self._frequencies[self._tx_random][0]
+        # write to RFM channel registers...
+        self._write_u8(_REG_FRF_MSB, self._rfm_msb)
+        self._write_u8(_REG_FRF_MID, self._rfm_mid)
+        self._write_u8(_REG_FRF_LSB, self._rfm_lsb)
+        # set RFM datarate
+        self._write_u8(_REG_FEI_LSB, self._sf)
+        self._write_u8(_REG_FEI_MSB, self._bw)
+        self._write_u8(_REG_MODEM_CONFIG, self._modemcfg)
+        # set RegPayloadLength
+        self._write_u8(_REG_PAYLOAD_LENGTH, packet_length)
+        # initalize FIFO pointer to base address for TX
+        self._write_u8(_REG_FIFO_POINTER, _REG_FIFO_BASE_ADDR)
+        # fill the FIFO buffer with the LoRa payload
+        k = 0  # ptr
+        i = 0
+        while i < packet_length:
+            self._write_u8(0x00, lora_packet[k])
+            i += 1
+            k += 1
+        # switch RFM to TX operating mode
+        self._write_u8(_REG_OPERATING_MODE, _MODE_TX)
+        # wait for TxDone IRQ
+        print('Sending packet')
+        while not self._irq.value:
+            # waiting for TxDone
+            print('.', end='')
```

Comment:
Should there be a timeout here? When testing is done - the print should be removed.

Reply:
Yep, added in a 15sec. timeout.